### PR TITLE
added filter for interactable press receiver and updated hololens 2 b…

### PR DESCRIPTION
**Pull request by thalbern**
## Overview
Interactable press receiver can now filter for far, near, or both event types. This allows us to properly configure buttons so they play sounds only from the front.

Also includes a fix for our InspectorField serialization whenever we're serializing newly added fields that aren't part of existing data/assets.

## Changes
- Fixes: #5616, #5647

### CODE REVIEW

**Comment by cre8ivepark** [comment]
Tested on HL2 device with 'Far only' filter option in the Interactable. It works well!
Also updated the blue Round button prefab as well.

2019-08-13 16_05_21-Unity 2018 4 5f1 Personal - PressableButtonExample unity - MRTK-Public-Microsoft

**Comment by cre8ivepark** [comment]
Also tested on HL1 and WMR headset.

**Comment by julenka** [comment]
I'm okay with this, but it is not ideal. Would love to hear from @Troy-Ferrell @killerantz  or @Railboy  if we could have a better approach for serializing these fields.

**Comment by julenka** [comment]
I reset my feedback since I haven't tested thoroughly enough. @cre8ivepark  or @thalbern  are you able to test and verify on HL2? We should do that before checking in.

**Comment by Troy-Ferrell** [comment]
> I'm okay with this, but it is not ideal. Would love to hear from @Troy-Ferrell @killerantz  or @Railboy  if we could have a better approach for serializing these fields.

What do you mean serializing these fields?

**Comment by cre8ivepark** [comment]
@julenka I'll do testing for HL2 & HL1

**Comment by cre8ivepark** [comment]
Tested on HL2 and HL1. Looks good!

**Comment by thalbern** [comment]
did another test on HL2 and looks good to me

**Comment by thalbern** [comment]
upgrade tests failed for me. using existing assets and updating to new code will break the existing pressreceivers - they need to be  manually removed and readded

**Comment by thalbern** [comment]
did several other upgrade scenarios which all failed for me. don't merge this change!

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableOnPressReceiver.cs [hunk 1]
```diff
@@ -11,33 +12,71 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
         public UnityEvent OnRelease = new UnityEvent();
 
+        public enum PressType
+        {
+            NearAndFar = 0,
+            NearOnly = 1,
+            FarOnly = 2
+        }
+
+        [InspectorField(Label = "Press Interaction Type", Tooltip = "Specify whether press event is for near or far interaction", Type = InspectorField.FieldTypes.DropdownInt, Options = new string[] { "Near and Far", "Near Only", "Far Only" })]
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableOnPressReceiver.cs]: I am not a fan of the name "Press Interaction Type" Suggestions:

* Near / Far Interaction Filter
* Near or Far Interaction?
* Interaction Filter
* Interaction Distance Filter
* Event Filter Type
* Event Filter

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableOnPressReceiver.cs [hunk 2]
```diff
@@ -11,33 +12,71 @@ namespace Microsoft.MixedReality.Toolkit.UI
...
+        public int PressTypeFilter = (int)PressType.NearAndFar;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableOnPressReceiver.cs]: @Troy-Ferrell and @Railboy  berni + i couldn't find a way to actually use an enum for the public field, it looks like the custom inspector only supports int lists. Do you have ideas / suggestions for how to use an enum here?

File: Assets/MixedRealityToolkit/Utilities/InspectorFields/InspectorGenericFields.cs [hunk 3]
```diff
@@ -31,7 +31,10 @@ public static void LoadSettings(T target, List<InspectorPropertySetting> setting
                 foreach (var attr in attrs)
                 {
                     object value = InspectorField.GetSettingValue(settings, propInfo.Name);
-                    propInfo.SetValue(target, value);
+                    if (value != null)
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Utilities/InspectorFields/InspectorGenericFields.cs]: what happens in the null case? Does a default value fill in?

Comment (thalbern) [reply Assets/MixedRealityToolkit/Utilities/InspectorFields/InspectorGenericFields.cs]: afais it's just taking whatever default is defined in the class - but i'll check again if that's really the case

**Comment by julenka** [comment]
Verified that I can have two receivers: a near and a far, and do different things based on whether I pressed near and far:

image

File: Assets/MixedRealityToolkit/Utilities/InspectorFields/InspectorGenericFields.cs (hunk 3 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit/Utilities/InspectorFields/InspectorGenericFields.cs]: yeah confirmed. it's getting the defaults from the class and this is just overwriting whatever default value is already in there

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableOnPressReceiver.cs (hunk 1 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Events/InteractableOnPressReceiver.cs]: I vote for "Interaction Filter"

**Comment by thalbern** [comment]
did some more testing on HL2 and upgrade tests with old buttons - looks good to me now.

**Comment by cre8ivepark** [comment]
Doing upgrade testing with Periodic Table app now.

**Comment by thalbern** [comment]
> Finished upgrade test with Periodic Table app. Looks good!

thanks Yoon! :)

**Pull request merged**